// sidebars/examples/express.js
"use strict";

/*
 * Module dependencies.
 */
var express         = require('express');
var connectTimeout  = require('connect-timeout');
var sidebars        = require('../index');
var app             = module.exports = express();


/*
 * Configuration
 */
app.configure(function () {

  var timeout = connectTimeout({ time: 5000 });
  app.use(timeout);
  app.use( app.router );

  app.use(function (err, req, res, next) {
    console.log(err.stack);
    res.send(500, 'Something broke!');
  });
});

/*
 * Routes
 */
app.get( '/', function (req, res, next) {

  sidebars
    .engine( sidebars.handlebars )
    .use( sidebars.debug )
    .render(req, res, next, {
      template: 'hello.br',
      sidebars: {
        body: [
          {
            package : '../examples/widget',
            params : {
              id : 'one',
              content : 'Widget One'
            }
          },
          {
            package : '../examples/widget',
            params : {
              id : 'two',
              content : 'Widget Two'
            }
          }
        ],
        header: [
          {
            package : '../examples/widget',
            params : {
              id : 'header',
              content : 'Header Goes Here'
            }
          }
        ],
        footer: [
          {
            package : '../examples/widget',
            params : {
              id : 'footer',
              content : 'Footer Goes Here'
            }
          },
          {
            package : '../examples/widget',
            params : {
              id : 'subfooter',
              content : 'Sub Footer Goes Here'
            }
          }
        ],
      }
    });
});


app.listen(6000);
